gk/fib.c

```diff
@@ -1664,3 +1665,303 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	rte_memcpy(&dentry->grantor_ip, &fib->u.grantor.gt_addr,
+		sizeof(dentry->grantor_ip));
+
+	fillup_gk_fib_ether_entry(fib->u.grantor.eth_cache, eth_tbl);
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gateway_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	fillup_gk_fib_ether_entry(fib->u.gateway.eth_cache, eth_tbl);
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_neighbor_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int num_entries = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	struct neighbor_hash_table *neigh_ht;
+	struct gk_fib_ether_entry *eth_tbl;
+	struct gk_fib_ether_entry *entry;
+
+	if (dentry->addr.proto == ETHER_TYPE_IPv4)
+		neigh_ht = &fib->u.neigh;
+	else
+		neigh_ht = &fib->u.neigh6;
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		fillup_gk_fib_ether_entry(eth_cache, entry);
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+int
+print_fib_dump_entry(void *fib_dump_entry, __attribute__((unused)) void *arg)
+{
+	int ret;
+	int i = 0;
+	char ip_addr[128];
+	struct gk_fib_dump_entry *dentry = fib_dump_entry;
+
+	ret = convert_ip_to_str(&dentry->addr, ip_addr, sizeof(ip_addr));
+	if (ret < 0)
+		return -1;
+
+	RTE_LOG(INFO, GATEKEEPER,
+		"FIB entry for IP prefix: %s/%d with action %d.\t",
+		ip_addr, dentry->len, dentry->action);
```

Comment:
We shouldn't output to the log of Gatekeeper but to the output of the Lua script requesting it, so `gkctl` will get the output.

Notice that this comment applies to other calls of `RTE_LOG()` in `print_fib_dump_entry()`.